kubearmor/KubeArmor
# feat(operator): kubearmor-operator

**Purpose of PR?**:
This PR is continuation of existing PR #925 , it adds KubeArmor Operator, it's a basic installation operator that can deploy KubeArmor on any k8s cluster considering the dynamic node configurations in terms of container runtime and LSM ( Linux Security Module). 

**Changes Includes**

- Created a CRD `KubeArmorConfig` to define configurations options for the KubeArmor.
- Implemented reconciler/watcher to deploy the KubeArmor applications with desire state.
- Added CI workflow to test and publish KubeArmor Operator to the docker registry `kubearmor/kubearmor-operator`.

**How to Test**
- Build a local image of the kubearmor-operator using `make docker build`
- deploy the operator using either helm chart or deployment file `deployments/operator.yaml` and use the locally build image there.
- after operator got deployed and running, apply sample (kubearmorconfig) crd `config/samples/sample-config.yml`
- observe the crd status and once it shows status `Running` it means all the required KubeArmor Components i.e., daemonset, kubearmor-relay, kubearmor-controller, etc. has been deployed on the cluster.
- try updating the `kubearmorconfig` and verify that KubeArmor Configured accordingly.

```
$ kubectl get pods -A
NAMESPACE     NAME                                  READY   STATUS    RESTARTS   AGE
kube-system   aws-node-b9vsm                        1/1     Running   0          73d
kube-system   aws-node-hpj66                        1/1     Running   0          73d
kube-system   coredns-7975d6fb9b-5v948              1/1     Running   0          73d
kube-system   coredns-7975d6fb9b-dzbxj              1/1     Running   0          73d
kube-system   kube-proxy-ftgzd                      1/1     Running   0          73d
kube-system   kube-proxy-n26q6                      1/1     Running   0          73d
kube-system   kubearmor-operator-6fd4b5d465-qrzlk   1/1     Running   0          7s

$ kubectl get kubearmorconfig -A -w
NAMESPACE     NAME                     STATUS
kube-system   kubearmorconfig-sample   Pending
kube-system   kubearmorconfig-sample   Running

$ kubectl get pods -A
kube-system   aws-node-b9vsm                          1/1     Running   0          73d
kube-system   aws-node-hpj66                          1/1     Running   0          73d
kube-system   coredns-7975d6fb9b-5v948                1/1     Running   0          73d
kube-system   coredns-7975d6fb9b-dzbxj                1/1     Running   0          73d
kube-system   kube-proxy-ftgzd                        1/1     Running   0          73d
kube-system   kube-proxy-n26q6                        1/1     Running   0          73d
kube-system   kubearmor-bpf-containerd-98c2c-7tdk8    1/1     Running   0          2m5s
kube-system   kubearmor-bpf-containerd-98c2c-9r7b2    1/1     Running   0          2m5s
kube-system   kubearmor-controller-7c48c74d58-n7s99   2/2     Running   0          2m3s
kube-system   kubearmor-operator-6fd4b5d465-qrzlk     1/1     Running   0          4m17s
kube-system   kubearmor-relay-6fddb8865b-86hfx        1/1     Running   0          2m42s
```

**HLD Diagram**

The linked issue #914 includes an attached High-Level Design (HLD) diagram illustrating the design of the KubeArmor Operator.

Attaching another diagram here that illustrates kubearmor deployment workflow using the operator:
![kubearmor-operator](https://github.com/kubearmor/KubeArmor/assets/81104646/e57b155c-0a56-4fa5-809c-5802af95cb1c)

**Does this PR introduce a breaking change?**
No
**If the changes in this PR are manually verified, list down the scenarios covered:**:

**Additional information for reviewer?** :
_Mention if this PR is part of any design or a continuation of previous PRs_


**Checklist:**
- [x] PR Title follows the convention of  `<type>(<scope>): <subject>`
